### PR TITLE
CodeToCloud tool feature parity for Deploy and Quota area

### PR DESCRIPTION
**Pull request by wchigit**
## What does this PR do?
This PR updates tools in Deploy and Quota areas to meet the latest feature improvements from CodeToCloud.
Main features include:
- Support deployment using Azure CLI+Bicep/Terraform
- Support deployment to existing resources
- Add Azure-specific IaC rules regarding code quality, configuration success rate and security
- Update CI/CD pipeline generation user flow

`[Any additional context, screenshots, or information that helps reviewers]`

## GitHub issue number?
`[Link to the GitHub issue this PR addresses]`

## Pre-merge Checklist
- [ ] Required for All PRs
    - [ ] **Read [contribution guidelines](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md)**
    - [ ] PR title clearly describes the change
    - [ ] Commit history is clean with descriptive messages ([cleanup guide](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md))
    - [ ] Added comprehensive tests for new/modified functionality
    - [ ] Updated `servers/Azure.Mcp.Server/CHANGELOG.md` and/or `servers/Fabric.Mcp.Server/CHANGELOG.md` for product changes (`features, bug fixes, UI/UX, updated dependencies`)
- [ ] For MCP tool changes:
    - [ ] **One tool per PR**: This PR adds or modifies only one MCP tool for faster review cycles
    - [ ] Updated `servers/Azure.Mcp.Server/README.md` and/or `servers/Fabric.Mcp.Server/README.md` documentation
    - [ ] Validate README.md changes using script at `eng/scripts/Process-PackageReadMe.ps1`. See [Package README](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md#package-readme)
    - [ ] Updated command list in `/servers/Azure.Mcp.Server/docs/azmcp-commands.md` and/or `/docs/fabric-commands.md`
    - [ ] Run `.\eng\scripts\Update-AzCommandsMetadata.ps1` to update tool metadata in azmcp-commands.md (required for CI)
    - [ ] For new or modified tool descriptions, ran [`ToolDescriptionEvaluator`](https://github.com/microsoft/mcp/blob/main/eng/tools/ToolDescriptionEvaluator/Quickstart.md) and obtained a score of `0.4` or more and a top 3 ranking for all related test prompts
    - [ ] For tools with new names, including new tools or renamed tools, update [`consolidated-tools.json`](https://github.com/microsoft/mcp/blob/main/core/Azure.Mcp.Core/src/Areas/Server/Resources/consolidated-tools.json)
    - [ ] For new tools associated with Azure services or publicly available tools/APIs/products, add URL to documentation in the PR description
- [ ] Extra steps for **Azure MCP Server** tool changes:
    - [ ] Updated test prompts in `/servers/Azure.Mcp.Server/docs/e2eTestPrompts.md`
    - [ ] 👉 For Community (non-Microsoft team member) PRs:
        - [ ] **Security review**: Reviewed code for security vulnerabilities, malicious code, or suspicious activities before running tests (`crypto mining, spam, data exfiltration, etc.`)
        - [ ] **Manual tests run**: added comment `/azp run mcp - pullrequest - live` to run *Live Test Pipeline*

### CODE REVIEW

**Comment by wchigit** [comment]
@microsoft/azure-mcp Hi team, could you help review this PR?

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs [hunk 1]
```diff
@@ -22,12 +27,8 @@ public static class IaCRulesTemplateUtil
...
-        if (deploymentTool.Equals(DeploymentTool.AzCli, StringComparison.OrdinalIgnoreCase))
-        {
-            return TemplateService.LoadTemplate("IaCRules/azcli-rules");
-        }
         // Default values for optional parameters
-        if (string.IsNullOrWhiteSpace(iacType))
+        if (deploymentTool.Equals(DeploymentTool.Azd, StringComparison.OrdinalIgnoreCase) && string.IsNullOrWhiteSpace(iacType))
         {
             iacType = "bicep";
         }
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs:33]: This is good feedback from copilot. 
Also, add unit tests for azd + empty iacType case.

File: tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs:1]: This filename has a trailing whitespace, fix it.

File: tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker .cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            {
+                var name = string.Empty;
+
+                if (item.TryGetProperty("name", out var nameElement))
+                {
+                    name = nameElement.GetStringSafe();
+                }
+
+                // Filter by specific SKU names
+                if (!SkuNameList.Contains(name))
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker .cs]: This is case sensitive ordinal comparison.
Should use 
```c#
if (!SkuNameList.Contains(name, StringComparer.OrdinalIgnoreCase))
```

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs [hunk 4]
```diff
@@ -4,6 +4,9 @@
 using Azure.Mcp.Tools.Deploy.Models;
 using Azure.Mcp.Tools.Deploy.Models.Templates;
 using Azure.Mcp.Tools.Deploy.Services.Templates;
+using Microsoft.AspNetCore.DataProtection;
+using ModelContextProtocol.Protocol;
+using YamlDotNet.Core.Tokens;
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs]: This is valid feedback.

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/PipelineGenerationUtil.cs [hunk 5]
```diff
@@ -1,9 +1,18 @@
...
+using System.Security.Cryptography;
+using System.Text.RegularExpressions;
 using Azure.Mcp.Tools.Deploy.Models.Templates;
 using Azure.Mcp.Tools.Deploy.Options.Pipeline;
 using Azure.Mcp.Tools.Deploy.Services.Templates;
+using Azure.ResourceManager.Network.Models;
+using Microsoft.AspNetCore.Http.HttpResults;
+using OpenTelemetry;
+using static System.Net.Mime.MediaTypeNames;
+using static System.Runtime.InteropServices.JavaScript.JSType;
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply tools/Azure.Mcp.Tools.Deploy/src/Services/Util/PipelineGenerationUtil.cs]: This is valid feedback.

File: tools/Azure.Mcp.Tools.Deploy/src/Options/DeployOptionDefinitions.cs [hunk 6]
```diff
@@ -56,53 +56,49 @@ public class AzdAppLogOptions : SubscriptionOptions
 
     public class PipelineGenerateOptions : SubscriptionOptions
     {
-        public const string UseAZDPipelineConfigName = "use-azd-pipeline-config";
-        public const string OrganizationNameName = "organization-name";
-        public const string RepositoryNameName = "repository-name";
-        public const string GithubEnvironmentNameName = "github-environment-name";
+        public const string isAZDProjectName = "is-azd-project";
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Deploy/src/Options/DeployOptionDefinitions.cs]: camel case is not best practice in C#
use pascal case here and for other variable names in this file.

File: tools/Azure.Mcp.Tools.Deploy/src/Options/DeployOptionDefinitions.cs [hunk 7]
```diff
@@ -56,53 +56,49 @@ public class AzdAppLogOptions : SubscriptionOptions
...
     {
         public const string WorkspaceFolderName = "workspace-folder";
         public const string ProjectNameName = "project-name";
         public const string TargetAppServiceName = "target-app-service";
         public const string ProvisioningToolName = "provisioning-tool";
-        public const string AzdIacOptionsName = "azd-iac-options";
+        public const string IacOptionsName = "iac-options";
+        public const string SourceTypeName = "source-type";
+        public const string DeployOptionName = "deploy-option";
+        public const string ResourceGroupNameName = "resource-group-name";
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Deploy/src/Options/DeployOptionDefinitions.cs]: Resource Group option is commonly defined here.
OptionDefinitions.Common.ResourceGroup
We should ideally use that, instead of defining a duplicate here

**Review by qianwens — APPROVED** [review]

File: AzureMcp.sln [hunk 8]
```diff
@@ -0,0 +1,2403 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (g2vinay) [inline AzureMcp.sln]: The repo already uses Microsoft.Mcp.slnx This looks like an accidentally committed local file. Should be removed

File: tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+        return $"{name} (remaining quota: {remainingQuota})";
+    }
+
+    private static List<UsageInfo> CreateEmptyQuotaInfo() =>
+    [
+        new UsageInfo(Name: "No SKU available", Limit: 0, Used: 0)
+    ];
+
+    protected async Task<JsonDocument?> GetQuotaByUrlAsync(string requestUrl, CancellationToken cancellationToken = default)
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs]: This method exists in PostgresSQLUsageChecker.cs
instead of duplicating it, mov
This logic belongs in the AzureUsageChecker base class as a shared protected helper. 
And then ensure both the classes use it from there, instead of duplicating it.

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs [hunk 10]
```diff
@@ -139,29 +106,137 @@ private static string GenerateResourceSpecificRules(IaCRulesTemplateParameters p
...
         }
 
+        rules.Add(GenerateKeyVaultRules(parameters));
+
         return string.Join(Environment.NewLine, rules);
     }
 
+    private static string GetToolSpecificResourceRules(string iacType, string? bicepRules, string? tfRules, string? cliRules)
+    {
+        return iacType switch
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs]: iacType here is parameters.IacType, set directly from user input, it is not normalized before being passed in. If the user provides "Bicep" or "Terraform" (any casing other than all lowercase), the switch falls through to CLI rules. 

 normalize iacType at the top of GetToolSpecificResourceRules using .ToLowerInvariant().

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs [hunk 11]
```diff
@@ -12,6 +13,8 @@ namespace Azure.Mcp.Tools.Deploy.Services.Util;
 /// </summary>
 public static class IaCRulesTemplateUtil
 {
+    private static string _databaseCommonRules = TemplateService.LoadTemplate("IaCRules/database-common-rules");
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs]: Missing readonly, the field can be mutated by any code in the class.

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/PipelineGenerationUtil.cs [hunk 12]
```diff
@@ -19,55 +19,121 @@ public static class PipelineGenerationUtil
...
     /// <summary>
-    /// Checks if the provided string is a valid GUID format.
+    /// Creates pipeline template parameters from the provided options.
     /// </summary>
-    private static bool CheckGUIDFormat(string input)
+    private static PipelineTemplateParameters CreatePipelineParameters(GuidanceGetOptions options)
     {
-        return Guid.TryParse(input, out _);
+        options.PipelinePlatform ??= Models.PipelinePlatform.GitHubActions;
+        options.DeployOption ??= Models.DeployOption.DeployOnly;
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Deploy/src/Services/Util/PipelineGenerationUtil.cs]: This has the side effect of modifying the caller's options object. The binding result is immutable by convention in MCP commands. Its best to copy values into local variables instead.

File: tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/ComputeUsageChecker.cs [hunk 13]
```diff
@@ -9,31 +10,146 @@
...
+    }
+
     public override async Task<List<UsageInfo>> GetUsageForLocationAsync(string location, CancellationToken cancellationToken)
     {
         try
         {
-            var subscription = ResourceClient.GetSubscriptionResource(new ResourceIdentifier($"/subscriptions/{SubscriptionId}"));
-            var usages = subscription.GetUsagesAsync(location, cancellationToken);
+            var skus = ResourceClient.GetSubscriptionResource(new ResourceIdentifier($"/subscriptions/{SubscriptionId}"))
+                .GetComputeResourceSkusAsync(cancellationToken: cancellationToken);
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/ComputeUsageChecker.cs]: Why are we not calling this with the location filter. 
The current code iterates the full global list and filters in-memory, which is significantly slower and increases API cost.
I believe, the Compute SKUs API supports server-side filtering, we should utilize it.

File: servers/Azure.Mcp.Server/docs/azmcp-commands.md [hunk 14]
```diff
@@ -1044,18 +1044,21 @@ azmcp deploy iac rules get --deployment-tool <deployment-tool> \
 
 # Get the ci/cd pipeline guidance
 # ❌ Destructive | ✅ Idempotent | ❌ OpenWorld | ✅ ReadOnly | ❌ Secret | ❌ LocalRequired
-azmcp deploy pipeline guidance get [--use-azd-pipeline-config <use-azd-pipeline-config>] \
-                                   [--organization-name <organization-name>] \
-                                   [--repository-name <repository-name>] \
-                                   [--github-environment-name <github-environment-name>]
+azmcp deploy pipeline guidance get [--is-azd-project <is-azd-project>] \
```
Comment (g2vinay) [inline servers/Azure.Mcp.Server/docs/azmcp-commands.md:1047]: We're changing multiple option names. 

--use-azd-pipeline-config → --is-azd-project
--organization-name, --repository-name, --github-environment-name → removed
--pipeline-platform, --deploy-option → new required options
--azd-iac-options → --iac-options

These are all breaking changes. 
Call them out as breaking changes in the changelog. 
Use the changelog-entries.md file in the repo to invoke the script that generates the changelog entry as .yaml file and check that file into this PR.

Also, include the changes being done in this PR in the changelog entry.

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs [hunk 15]
```diff
@@ -139,29 +106,137 @@ private static string GenerateResourceSpecificRules(IaCRulesTemplateParameters p
...
         }
 
-        if (parameters.ResourceTypes.Contains(AzureServiceNames.AzureStorage))
+        if (parameters.ResourceTypes.Contains(AzureServiceNames.AzureStorageAccount, StringComparer.OrdinalIgnoreCase))
         {
-            rules.Add(TemplateService.ProcessTemplate("IaCRules/storage-rules", parameters.ToDictionary()));
+            rules.Add(GenerateStorageRules(parameters));
         }
 
+        rules.Add(GenerateKeyVaultRules(parameters));
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs:149]: Why are adding KeyVault rules by default ?

Why the resourceTypes.Contains(AzureServiceNames.AzureKeyVault, ...) guard is missing.

File: tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        "SubnetQuota"
+    ];
+
+    private readonly IHttpClientFactory _httpClientFactory = httpClientFactory ?? throw new ArgumentNullException(nameof(httpClientFactory));
+
+    public override async Task<List<UsageInfo>> GetUsageForLocationAsync(string location, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var apiVersion = "2015-05-01-preview";
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs]: This is a 10-year-old preview API. Even if the endpoint is still available, the version should be surfaced as a named constant, not an inline magic string.

Also, check and use the latest stable API version.

File: tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs (hunk 15 above)
Comment (wchigit) [reply tools/Azure.Mcp.Tools.Deploy/src/Services/Util/IaCRulesTemplateUtil.cs:149]: This is by design. Key vault is not a service that Copilot can easily detect from project while it is required service for Azure security.

File: tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs (hunk 16 above)
Comment (wchigit) [reply tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/SQLUsageChecker.cs]: updated

File: tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/ComputeUsageChecker.cs (hunk 13 above)
Comment (wchigit) [reply tools/Azure.Mcp.Tools.Quota/src/Services/Util/Usage/ComputeUsageChecker.cs]: region filter added
